src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -28,11 +28,37 @@ public static class BuildCallGraph
         public static CallGraph Apply(QsCompilation compilation)
         {
             var walker = new BuildGraph();
-            foreach (var ns in compilation.Namespaces)
+
+            if (compilation.EntryPoints.Any())
             {
-                walker.Namespaces.OnNamespace(ns);
+                walker.SharedState.IsLimitedToEntryPoints = true;
+                walker.SharedState.RequestStack = new Stack<QsQualifiedName>(compilation.EntryPoints);
+                walker.SharedState.ResolvedCallableSet = new HashSet<QsQualifiedName>();
+                var globals = compilation.Namespaces.GlobalCallableResolutions();
+                while (walker.SharedState.RequestStack.Any())
+                {
+                    var currentRequest = walker.SharedState.RequestStack.Pop();
+
+                    // If there is a call to an unknown callable, throw exception
+                    if (!globals.TryGetValue(currentRequest, out QsCallable currentCallable))
+                        throw new ArgumentException($"Couldn't find definition for callable: {currentRequest}");
+
+                    // The current request must be added before it is processed to prevent
+                    // self-references from duplicating on the stack.
+                    walker.SharedState.ResolvedCallableSet.Add(currentRequest);
+
+                    walker.Namespaces.OnCallableDeclaration(currentCallable);
+                }
             }
-            return walker.SharedState.graph;
+            else
+            {
```

Comment:
code inside `then/else` clauses of this `if` statement differs drastically and is kind of at different level of abstraction. It would potentially create a breeding ground for bugs: what if `OnNamespace` does something that is relevant to the subtree of entrypoints as well? Maybe, not right now, but what if someone adds such functionality in the future -- would they remember to update the code in `then` clause as well?

Could the driving logic of the two graph building strategies be more unified?

Reply:
I will take another look.

Reply:
(Sorry for the long comment) I have two thoughts on this: 1. My general thinking on concerns of the form "If someone changes X, then they will also have to change Y" is that if someone is changing things, then they will have to be responsible for make that change correctly and adjusting all the appropriate code. We can try to anticipate likely future changes and try to structure the code such that that future change is made easier and less error prone, but we can't plan for all possible future changes. Also planning for future changes that are not likely can make our code base worse, as complicating our code in preparation for a change that will never happen can make it more difficult to implement changes that *do* happen. So my general thinking on these kinds of concerns is: let's keep it simple for now and cross that bridge when we get to it.

Reply:
2. My second thought is that this method does two different things depending on a piece of information found inside the data structure of its argument, which is probably bad design. I think, rather than trying to unify this logic, it would make more sense to just have two separate methods for using this walker: one for libraries, and another for executables.

Reply:
This was my main concern: the logic in the two branches of the `if` statement is too different and essentially unrelated so it makes the code hard to maintain correctly. If you can handle these two cases in a more encapsulated way, I believe that would improve maintainability of the code.

As for the philosophical discussion re "If someone changes X, then they will also have to change Y". The question is, how will our future selves discover that they need to change Y when they embark on changing X? Does the architecture of the codebase guides them towards a complete change? Does a reasonable search for references to X finds Y so the references can be reviewed and updated if needed? Does the test coverage help with the discoverability of the related bits that must be changed together?  


---
In reply to: [441202023](https://github.com/microsoft/qsharp-compiler/pull/454#discussion_r441202023) [](ancestors = 441202023)

Reply:
I've separated the two branches out into separate methods. The original Apply method now just contains the discrimination logic between the two cases.